File: src/run/steps/pane.ts
import { HerdrError } from "../../host";
import type { PaneOpen } from "../../workflow/types";
import { sizeToFirstRatio, type RunnerDeps } from "../context";

export type PlacedPane = { pane_id: string; tab_id: string; workspace_id: string };

type PlaceAnchors = { paneId?: string; tabId?: string; workspaceId?: string };

export type PlaceOpts = {
  open: PaneOpen;
  target?: string;
  workspace?: string;
  size?: number;
  focus: boolean;
  cwd?: string;
  env?: Record<string, string>;
  label?: string;
  deps: { herdrCall: RunnerDeps["herdrCall"] };
  invocation: PlaceAnchors;
};

type PaneInfoish = { pane_id?: unknown; tab_id?: unknown; workspace_id?: unknown };

function fail(detail: string): never {
  throw new HerdrError("placement_failed", detail);
}

function requireWorkspace(o: PlaceOpts): string {
  const workspace = o.workspace ?? o.invocation.workspaceId;
  if (!workspace) fail("pane.open: tab needs pane.workspace or an invocation workspace");
  return workspace;
}

function requireTargetPane(o: PlaceOpts): string {
  const target = o.target ?? o.invocation.paneId;
  if (!target) fail(`pane.open: ${o.open} needs pane.target or an invocation pane`);
  return target;
}

function splitDirection(open: PaneOpen): "right" | "down" {
  return open === "beside" ? "right" : "down";
}

function placedFrom(source: unknown, where: string): PlacedPane {
  const info = (source ?? {}) as PaneInfoish;
  const paneId = info.pane_id;
  const tabId = info.tab_id;
  const workspaceId = info.workspace_id;
  if (typeof paneId !== "string" || typeof tabId !== "string" || typeof workspaceId !== "string") {
    fail(`${where} did not return pane/tab/workspace identifiers`);
  }
  return { pane_id: paneId, tab_id: tabId, workspace_id: workspaceId };
}

/** Empty shell pane for a managed agent: tab.create for tabs, pane.split for beside/below. */
export async function placeEmptyPane(o: PlaceOpts): Promise<PlacedPane> {
  if (o.open === "tab") {
    const result = await o.deps.herdrCall("tab.create", {
      workspace_id: requireWorkspace(o),
      cwd: o.cwd ?? null,
      env: o.env ?? {},
      focus: o.focus,
      label: o.label ?? null,
    });
    const tab = (result.tab ?? {}) as PaneInfoish;
    const pane = placedFrom(result.root_pane, "tab.create");
    return {
      pane_id: pane.pane_id,
      tab_id: typeof tab.tab_id === "string" ? tab.tab_id : pane.tab_id,
      workspace_id: typeof tab.workspace_id === "string" ? tab.workspace_id : pane.workspace_id,
    };
  }
  const result = await o.deps.herdrCall("pane.split", {
    direction: splitDirection(o.open),
    target_pane_id: requireTargetPane(o),
    ratio: o.size !== undefined ? sizeToFirstRatio(o.size) : null,
    cwd: o.cwd ?? null,
    env: o.env ?? {},
    focus: o.focus,
  });
  return placedFrom(result.pane, "pane.split");
}

type LayoutNodeish = {
  pane_id?: unknown;
  second?: LayoutNodeish;
};

type LayoutResult = {
  tab_id?: unknown;
  workspace_id?: unknown;
  focused_pane_id?: unknown;
  root?: LayoutNodeish;
};

function createdPaneId(layout: LayoutResult, split: boolean): string {
  const node = split ? layout.root?.second : layout.root;
  const fromTree = node?.pane_id;
  if (typeof fromTree === "string") return fromTree;
  if (typeof layout.focused_pane_id === "string") return layout.focused_pane_id;
  return fail("layout.apply did not return the created pane id");
}

function layoutPlacement(result: Record<string, unknown>, split: boolean): PlacedPane {
  const layout = (result.layout ?? {}) as LayoutResult;
  const tabId = layout.tab_id;
  const workspaceId = layout.workspace_id;
  if (typeof tabId !== "string" || typeof workspaceId !== "string") {
    fail("layout.apply did not return tab/workspace identifiers");
  }
  return { pane_id: createdPaneId(layout, split), tab_id: tabId, workspace_id: workspaceId };
}

/** Quote argv for submission into an interactive shell via pane.send_input. */
export function quoteArgvForShell(argv: string[]): string {
  return argv.map(quotePosixArg).join(" ");
}

function quotePosixArg(value: string): string {
  if (value.length === 0) return "''";
  if (/^[A-Za-z0-9_./:=+-]+$/.test(value)) return value;
  return `'${value.replace(/'/g, `'\\''`)}'`;
}

/**
 * Pane running a command.
 * - `open: tab` uses layout.apply with a command leaf (true argv, no shell).
 * - `open: beside|below` uses pane.split (preserves the anchor process) then
 *   pane.send_input with a shell-quoted command line — Herdr has no pane.run
 *   socket method, and layout.apply replaces the tab without preserving PTYs.
 */
export async function placeCommandPane(o: PlaceOpts & { argv: string[] }): Promise<PlacedPane> {
  if (o.open === "tab") {
    const result = await o.deps.herdrCall("layout.apply", {
      workspace_id: requireWorkspace(o),
      tab_label: o.label ?? null,
      tab_id: null,
      focus: o.focus,
      root: {
        type: "pane",
        label: o.label ?? null,
        cwd: o.cwd ?? null,
        command: o.argv,
        env: o.env ?? {},
      },
    });
    return layoutPlacement(result, false);
  }
  const placed = await placeEmptyPane(o);
  await o.deps.herdrCall("pane.send_input", {
    pane_id: placed.pane_id,
    text: quoteArgvForShell(o.argv),
    keys: ["Enter"],
  });
  return placed;
}

File: internal/engine/pane.go
package engine

import (
	"fmt"
	"regexp"
	"slices"
	"strings"

	"github.com/aorumbayev/herdr-workflows/internal/config"
	"github.com/aorumbayev/herdr-workflows/internal/host"
	"github.com/aorumbayev/herdr-workflows/internal/workflow"
)

// PlacedPane is the pane/tab/workspace identifier triple.
type PlacedPane = workflow.ReadinessResult

// PlaceOpts specifies pane placement parameters.
type PlaceOpts struct {
	Open       string
	Anchor     string
	Workspace  string
	Size       *int
	Focus      bool
	Cwd        string
	Env        map[string]string
	Label      string
	Argv       []string
	Deps       RunnerDeps
	Invocation config.InvocationContext
}

func nullableStr(s string) any {
	if s == "" {
		return nil
	}
	return s
}

func SizeToFirstRatio(sizePercent int) float64 {
	return float64(100-sizePercent) / 100.0
}

func failPlacement(detail string) error {
	return &host.HerdrError{Code: "placement_failed", Msg: detail}
}

func requireWorkspace(o PlaceOpts) (string, error) {
	workspace := o.Workspace
	if workspace == "" {
		workspace = o.Invocation.WorkspaceID
	}
	if workspace == "" {
		return "", failPlacement("pane.open: tab needs pane.workspace or an invocation workspace")
	}
	return workspace, nil
}

func requireAnchorPane(o PlaceOpts) (string, error) {
	anchor := o.Anchor
	if anchor == "" {
		anchor = o.Invocation.PaneID
	}
	if anchor == "" {
		return "", failPlacement(fmt.Sprintf("pane.open: %s needs pane.target or an invocation pane", o.Open))
	}
	return anchor, nil
}

func splitDirection(open string) string {
	if open == "beside" {
		return "right"
	}
	return "down"
}

func placedFrom(source any, where string) (PlacedPane, error) {
	m, ok := source.(map[string]any)
	if !ok {
		m = map[string]any{}
	}

	paneID, _ := m["pane_id"].(string)
	tabID, _ := m["tab_id"].(string)
	workspaceID, _ := m["workspace_id"].(string)

	if paneID == "" || tabID == "" || workspaceID == "" {
		return PlacedPane{}, failPlacement(fmt.Sprintf("%s did not return pane/tab/workspace identifiers", where))
	}

	return PlacedPane{
		PaneID:      paneID,
		TabID:       tabID,
		WorkspaceID: workspaceID,
	}, nil
}

func placeEmptyTab(o PlaceOpts) (PlacedPane, error) {
	workspace, err := requireWorkspace(o)
	if err != nil {
		return PlacedPane{}, err
	}

	env := o.Env
	if env == nil {
		env = map[string]string{}
	}

	params := map[string]any{
		"workspace_id": workspace,
		"cwd":          nullableStr(o.Cwd),
		"env":          env,
		"focus":        o.Focus,
		"label":        nullableStr(o.Label),
	}

	result, err := o.Deps.HerdrCall("tab.create", params)
	if err != nil {
		return PlacedPane{}, err
	}

	tab, _ := result["tab"].(map[string]any)
	rootPane, _ := result["root_pane"].(map[string]any)

	pane, err := placedFrom(rootPane, "tab.create")
	if err != nil {
		return PlacedPane{}, err
	}

	tabID := pane.TabID
	workspaceID := pane.WorkspaceID
	if tabIDFromTab, ok := tab["tab_id"].(string); ok {
		tabID = tabIDFromTab
	}
	if workspaceIDFromTab, ok := tab["workspace_id"].(string); ok {
		workspaceID = workspaceIDFromTab
	}

	return PlacedPane{
		PaneID:      pane.PaneID,
		TabID:       tabID,
		WorkspaceID: workspaceID,
	}, nil
}

// PlaceEmptyPane creates an empty pane for a managed agent.
func PlaceEmptyPane(o PlaceOpts) (PlacedPane, error) {
	if o.Open == "tab" {
		return placeEmptyTab(o)
	}

	anchor, err := requireAnchorPane(o)
	if err != nil {
		return PlacedPane{}, err
	}

	env := o.Env
	if env == nil {
		env = map[string]string{}
	}

	params := map[string]any{
		"direction":      splitDirection(o.Open),
		"target_pane_id": anchor,
		"ratio":          nil,
		"cwd":            nullableStr(o.Cwd),
		"env":            env,
		"focus":          o.Focus,
	}

	if o.Size != nil {
		params["ratio"] = SizeToFirstRatio(*o.Size)
	}

	result, err := o.Deps.HerdrCall("pane.split", params)
	if err != nil {
		return PlacedPane{}, err
	}

	pane, _ := result["pane"].(map[string]any)
	return placedFrom(pane, "pane.split")
}

// PlaceCommandPane creates a pane running a command.
func PlaceCommandPane(o PlaceOpts) (PlacedPane, error) {
	if o.Open == "tab" {
		workspace, err := requireWorkspace(o)
		if err != nil {
			return PlacedPane{}, err
		}

		env := o.Env
		if env == nil {
			env = map[string]string{}
		}

		params := map[string]any{
			"workspace_id": workspace,
			"tab_label":    nullableStr(o.Label),
			"tab_id":       nil,
			"focus":        o.Focus,
			"root": map[string]any{
				"type":    "pane",
				"label":   nullableStr(o.Label),
				"cwd":     nullableStr(o.Cwd),
				"command": o.Argv,
				"env":     env,
			},
		}

		result, err := o.Deps.HerdrCall("layout.apply", params)
		if err != nil {
			return PlacedPane{}, err
		}

		return createdPaneFromLayout(result, false)
	}

	placed, err := PlaceEmptyPane(o)
	if err != nil {
		return PlacedPane{}, err
	}

	text := QuoteArgvForShell(o.Argv)
	params := map[string]any{
		"pane_id": placed.PaneID,
		"text":    text,
		"keys":    []string{"Enter"},
	}

	_, err = o.Deps.HerdrCall("pane.send_input", params)
	if err != nil {
		return PlacedPane{}, err
	}

	return placed, nil
}

func createdPaneFromLayout(result map[string]any, split bool) (PlacedPane, error) {
	layout, _ := result["layout"].(map[string]any)

	tabID, _ := layout["tab_id"].(string)
	workspaceID, _ := layout["workspace_id"].(string)

	if tabID == "" || workspaceID == "" {
		return PlacedPane{}, failPlacement("layout.apply did not return tab/workspace identifiers")
	}

	paneID, err := createdPaneID(layout, split)
	if err != nil {
		return PlacedPane{}, err
	}

	return PlacedPane{
		PaneID:      paneID,
		TabID:       tabID,
		WorkspaceID: workspaceID,
	}, nil
}

func createdPaneID(layout map[string]any, split bool) (string, error) {
	var node map[string]any
	if split {
		root, _ := layout["root"].(map[string]any)
		second, _ := root["second"].(map[string]any)
		node = second
	} else {
		root, _ := layout["root"].(map[string]any)
		node = root
	}

	if paneID, ok := node["pane_id"].(string); ok && paneID != "" {
		return paneID, nil
	}

	if focusedID, ok := layout["focused_pane_id"].(string); ok && focusedID != "" {
		return focusedID, nil
	}

	return "", failPlacement("layout.apply did not return the created pane id")
}

// QuoteArgvForShell quotes a command array for shell execution.
func QuoteArgvForShell(argv []string) string {
	parts := make([]string, len(argv))
	for i, arg := range argv {
		parts[i] = QuotePosixArg(arg)
	}
	return strings.Join(parts, " ")
}

// QuotePosixArg quotes a single shell argument for POSIX shells.
func QuotePosixArg(value string) string {
	if value == "" {
		return "''"
	}

	// regex: ^[A-Za-z0-9_./:=+-]+$
	isSafe := regexp.MustCompile(`^[A-Za-z0-9_./:=+-]+$`).MatchString(value)
	if isSafe {
		return value
	}

	// Single-quote with escaped internal single quotes
	return "'" + strings.ReplaceAll(value, "'", "'\\''") + "'"
}

// ResolvePaneLabel resolves a pane name template, falling back to the step ID if blank.
func ResolvePaneLabel(name string, ns workflow.TemplateNamespace, fallback string) string {
	if name == "" {
		return fallback
	}

	rendered := workflow.SubstituteText(name, ns)
	rendered = strings.TrimSpace(rendered)
	if rendered == "" {
		return fallback
	}

	return rendered
}

// ResolvePaneOpen validates and resolves a pane.open value.
func ResolvePaneOpen(open string, ns workflow.TemplateNamespace) (string, error) {
	if slices.Contains(workflow.PaneOpens, open) {
		return open, nil
	}

	resolved := workflow.SubstituteValue(open, ns)
	resolvedStr := workflow.RenderScalar(resolved)

	if slices.Contains(workflow.PaneOpens, resolvedStr) {
		return resolvedStr, nil
	}

	return "", fmt.Errorf("pane.open resolved to '%s' (expected tab, beside, or below)", resolvedStr)
}
